### ISSUE TITLE
A small bug at initialization (Chapter 19, Subsection 19.3.2)

### ISSUE THREAD

**Issue by irudnyts** [hadley/mastering-shiny#481]
TL;DR: At initialization, the module server returns `NULL`, as long as `choices` of `selectInput()` is set to `NULL`. 

Consider the following module: 

```
selectVarInput <- function(id) {
  selectInput(NS(id, "var"), "Variable", choices = NULL) 
}

selectVarServer <- function(id, data, filter = is.numeric) {
  moduleServer(id, function(input, output, session) {
    observeEvent(data(), {
      updateSelectInput(session, "var", choices = find_vars(data(), filter))
    })
    
    reactive(data()[[input$var]])
  })
}
```

At initialization, first, `input$var` has `""` value, and the module `selectVarServer()` returns respectively `NULL` (data()[[""]]). Only after a moment, `observeEvent()` updates choices of the `input$var, and picks by default the first one. In this example, it is not critical, since `renderPrint()` can handle `NULL`. But if we would have to render a plot, for instance, the error would be displayed for a sec. See this reproducible example below: 

```
library(tidyverse)

datasetInput <- function(id, filter = NULL) {
  names <- ls("package:datasets")
  if (!is.null(filter)) {
    data <- lapply(names, get, "package:datasets")
    names <- names[vapply(data, filter, logical(1))]
  }
  
  selectInput(NS(id, "dataset"), "Pick a dataset", choices = names)
}

datasetServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    reactive(get(input$dataset, "package:datasets"))
  })
}

selectVarInput <- function(id) {
  selectInput(NS(id, "var"), "Variable", choices = NULL) 
}

find_vars <- function(data, filter) {
  names(data)[vapply(data, filter, logical(1))]
}

selectVarServer <- function(id, data, filter = is.numeric) {
  moduleServer(id, function(input, output, session) {
    observeEvent(data(), {
      updateSelectInput(session, "var", choices = find_vars(data(), filter))
    })
    
    reactive(data()[[input$var]])
  })
}

selectVarApp <- function(filter = is.numeric) {
  ui <- fluidPage(
    datasetInput("data", is.data.frame),
    selectVarInput("var"),
    plotOutput("out")
  )
  server <- function(input, output, session) {
    data <- datasetServer("data")
    var <- selectVarServer("var", data, filter = filter)
    output$out <- renderPlot(
      var() %>% 
        as_tibble() %>% 
        ggplot() + 
        geom_histogram(aes(x = value))
    )
  }
  
  shinyApp(ui, server)
}

selectVarApp()
```

The easy fix I see is to provide default choices of the first dataset and set `observeEvent`'s `ignoreInit` to `TRUE`: 

```
library(tidyverse)

datasetInput <- function(id, filter = NULL) {
  names <- ls("package:datasets")
  if (!is.null(filter)) {
    data <- lapply(names, get, "package:datasets")
    names <- names[vapply(data, filter, logical(1))]
  }
  
  selectInput(NS(id, "dataset"), "Pick a dataset", choices = names)
}

datasetServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    reactive(get(input$dataset, "package:datasets"))
  })
}

# define default choices
default_var_choices <- names(airquality)

selectVarInput <- function(id) {
  # set default choices
  selectInput(NS(id, "var"), "Variable", choices = default_var_choices) 
}

find_vars <- function(data, filter) {
  names(data)[vapply(data, filter, logical(1))]
}

selectVarServer <- function(id, data, filter = is.numeric) {
  moduleServer(id, function(input, output, session) {
    
    # disable trigger at initialization
    observeEvent(data(), {
      updateSelectInput(session, "var", choices = find_vars(data(), filter)), 
      ignoreInit = TRUE
    })
    
    reactive(data()[[input$var]])
  })
}

selectVarApp <- function(filter = is.numeric) {
  ui <- fluidPage(
    datasetInput("data", is.data.frame),
    selectVarInput("var"),
    plotOutput("out")
  )
  server <- function(input, output, session) {
    data <- datasetServer("data")
    var <- selectVarServer("var", data, filter = filter)
    output$out <- renderPlot(
      var() %>% 
        as_tibble() %>% 
        ggplot() + 
        geom_histogram(aes(x = value))
    )
  }
  
  shinyApp(ui, server)
}

selectVarApp()
```

**Comment by irudnyts** [comment]
Well, otherwise, one can simply validate if `var()` is numeric: 

```
library(tidyverse)

datasetInput <- function(id, filter = NULL) {
  names <- ls("package:datasets")
  if (!is.null(filter)) {
    data <- lapply(names, get, "package:datasets")
    names <- names[vapply(data, filter, logical(1))]
  }
  
  selectInput(NS(id, "dataset"), "Pick a dataset", choices = names)
}

datasetServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    reactive(get(input$dataset, "package:datasets"))
  })
}

selectVarInput <- function(id) {
  selectInput(NS(id, "var"), "Variable", choices = NULL)
}

find_vars <- function(data, filter) {
  names(data)[vapply(data, filter, logical(1))]
}

selectVarServer <- function(id, data, filter = is.numeric) {
  moduleServer(id, function(input, output, session) {
    
    observeEvent(data(), {
      updateSelectInput(session, "var", choices = find_vars(data(), filter))
    })
    
    reactive(data()[[input$var]])
  })
}

selectVarApp <- function(filter = is.numeric) {
  ui <- fluidPage(
    datasetInput("data", is.data.frame),
    selectVarInput("var"),
    plotOutput("out")
  )
  server <- function(input, output, session) {
    data <- datasetServer("data")
    var <- selectVarServer("var", data, filter = filter)
    output$out <- renderPlot({
      # validate that `var()` is numeric 🤘
      validate(
        need(is.numeric(var()), "Something wrong!")
      )
      var() %>% 
        as_tibble() %>% 
        ggplot() + 
        geom_histogram(aes(x = value))
    })
  }
  
  shinyApp(ui, server)
}

selectVarApp()

```

**Comment by irudnyts** [comment]
This is the last one, I promise! After reading [this](https://shiny.rstudio.com/articles/req.html), it seems the easiest way to do it is to validate the `input$var` in `reactive()` via `req()`: 

```
library(tidyverse)

datasetInput <- function(id, filter = NULL) {
  names <- ls("package:datasets")
  if (!is.null(filter)) {
    data <- lapply(names, get, "package:datasets")
    names <- names[vapply(data, filter, logical(1))]
  }
  
  selectInput(NS(id, "dataset"), "Pick a dataset", choices = names)
}

datasetServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    reactive(get(input$dataset, "package:datasets"))
  })
}

selectVarInput <- function(id) {
  selectInput(NS(id, "var"), "Variable", choices = NULL)
}

find_vars <- function(data, filter) {
  names(data)[vapply(data, filter, logical(1))]
}

selectVarServer <- function(id, data, filter = is.numeric) {
  moduleServer(id, function(input, output, session) {
    
    observeEvent(data(), {
      updateSelectInput(session, "var", choices = find_vars(data(), filter))
    })
    
    # validate that `input$var` is truthy 🤘
    reactive(data()[[req(input$var)]])
  })
}

selectVarApp <- function(filter = is.numeric) {
  ui <- fluidPage(
    datasetInput("data", is.data.frame),
    selectVarInput("var"),
    plotOutput("out")
  )
  server <- function(input, output, session) {
    data <- datasetServer("data")
    var <- selectVarServer("var", data, filter = filter)
    output$out <- renderPlot({
      var() %>% 
        as_tibble() %>% 
        ggplot() + 
        geom_histogram(aes(x = value))
    })
  }
  
  shinyApp(ui, server)
}

selectVarApp()
```
